fix: extract tarball to temp directory on Windows

dsanders11 opened the pull request:
<!--
Thank you for your pull request. Please review the below requirements.

Contributor guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `npm install && npm test` passes
- [x] tests are included <!-- Bug fixes and new features should include tests -->
- [ ] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/googleapis/release-please#how-should-i-write-my-commits)

##### Description of change
<!-- Provide a description of the change -->

`node-tar` unlinks file paths before writing to them on Windows, which causes errors when there are parallel installs going on which are extracting the headers tarball to `devDir`. This currently leads to random corruption of the cache at `devDir` with missing files and hanging files with `<filename>.DELETE.<random>` names.

This PR fixes the situation by:
* Fixing usage of `node-tar` to watch for errors during extraction
* Extracting the headers tarball on Windows to a temp directory and then copies the files over
* Using the `exponential-backoff` package in case of collision (`EBUSY`) when copying the files to `devDir`
* Bumps `installVersion` to force clearing of existing cached versions which might be corrupt on Windows

StefanStojanovic commented on test/test-install.js:
```diff
@@ -44,3 +52,119 @@ test('EACCES retry once', async (t) => {
+  semver.prerelease(process.version) !== null ||
+  semver.satisfies(process.version, '<10')
+
+test('parallel installs (ensure=true)', async (t) => {
```
Nitpicking: All `parallel installs ...` tests are similar (only differ in `prog.opts`). Potentially they can be refactored to a parametrized function to avoid code duplication.

dsanders11 replied:
Done.

dsanders11 commented on lib/install.js:
```diff
@@ -136,71 +168,99 @@ async function install (fs, gyp, argv) {
-        throw new Error(k + ' local checksum ' + contentShasums[k] + ' not match remote ' + expectShasums[k])
+      // copy over the files from the temp tarball extract directory to devDir
+      if (tarExtractDir !== devDir) {
+        await copyDirectory(tarExtractDir, devDir)
```
> Is there a reason we wouldn't want to `rename` / `move` here instead of copy?

Since there are parallel installs happening, there's no way to atomically rename to `devDir`. That directory is created (if it doesn't already exist) [at the start of the install process to confirm permissions](https://github.com/nodejs/node-gyp/blob/6f3c2d3c6c0de0dbf8c7245f34c2e0b3eea53812/lib/install.js#L106), so a direct rename of `tarExtractDir` to `devDir` would fail. Any rename process would need to first rename `devDir`, which opens a race condition  where `devDir` disappears momentarily during an ongoing build, causing a failure. That was [my original approach](https://github.com/dsanders11/node-gyp/commit/b21e845dd3217a0c4e5461af36962a275736f693#diff-a021b62b61739461595e009ed3d8524495144ddba458d99f1eba19e2a2be7b74R247-R255) for fixing this issue, but I switched to the current implementation to close that race condition.

> This could still race condition if it's not atomic?

The `copyFile` approach stays similar to the existing implementation, with `node-tar` overwriting the existing files when there are parallel installs going on. The main difference is that on Windows `node-tar` always unlinks the destination file path before writing to it to avoid certain conditions (same path multiple times in an archive) which don't exist for this usage.

I've stress tested this change (on CI and locally with a real package) with 35 parallel installs and didn't trip over any race conditions in practice.

dsanders11 commented on lib/install.js:
```diff
@@ -98,6 +100,35 @@ async function install (fs, gyp, argv) {
     }
   }
 
+  async function copyDirectory (src, dest) {
```
Made a change in e85492ed0cfeb9ca2e26f0d7c7b79c90584f8a23.

StefanStojanovic reviewed (approved):
LGTM. Since there are already 2 PRs addressing GitHub Action issues, we should make sure https://github.com/nodejs/node-gyp/pull/2846/commits/6bb7cd6e4192419afc137f022d6f4e9f1998429a is removed from here before landing it. Ideally, we'd first land a GitHub Actions fix, then rebase this PR, remove the mentioned commit and push changes.

StefanStojanovic commented:
@rvagg @cclauss I'm changing tests from `tap` to `mocha` in https://github.com/nodejs/node-gyp/pull/2851. Since this PR adds some tests and it's already approved, what are your thoughts on landing it (without https://github.com/nodejs/node-gyp/pull/2846/commits/6bb7cd6e4192419afc137f022d6f4e9f1998429a) so I can change those tests in my PR as well?

cclauss commented:
I have no objection to these changes but I am not a Windows user so I lack the skills to be a reviewer.

StefanStojanovic commented:
Hey, @dsanders11 can you prepare your branch for merging by removing https://github.com/nodejs/node-gyp/pull/2846/commits/6bb7cd6e4192419afc137f022d6f4e9f1998429a and potentially addressing https://github.com/nodejs/node-gyp/pull/2846#discussion_r1199659497 and squashing some commits together (especially the ones addressing PR comments)?

dsanders11 commented on lib/install.js:
```diff
@@ -98,6 +100,41 @@ async function install (fs, gyp, argv) {
+    const entries = await fs.promises.readdir(src, { withFileTypes: true })
+    for (const entry of entries) {
+      if (entry.isDirectory()) {
+        await fs.promises.mkdir(path.join(dest, entry.name), { recursive: true })
```
Yep, you're right, removed.

dsanders11 commented:
@StefanStojanovic, done! Removed 6bb7cd6 and squashed it back down to the original three commits. For CI reference, 2bb914b3c645881c0debb0cc0346d72f8cfc201b was the last commit before removing 6bb7cd6, and it went green.
